### ISSUE TITLE
Tools for performing importance sampling

### ISSUE THREAD

**Issue by EthanMarx** [ML4GW/BBHNet#251]
Develop tools for incorporating importance sampling. See https://arxiv.org/pdf/1712.00482.pdf

I'm imagining a class `ImportanceSampler` that is initialized with `source` and `target` `Callables` who's `__call__` method takes a sample in the parameter space (e.g (m1, m2) pair) and returns the probability under the distribution. 

`ImportanceSampler` will have a method `calculate_weights` that takes samples in the parameter space and calculates the corresponding importance sampling weights (i.e. the ratio between `target` and `source`). 

```python
@dataclass
class ImportanceSampler:
    target: Callable
    source: Callable

    def calculate_weights(self, samples):
        return self.target(samples) / self.source(samples)

```

It might be useful to find ways to integrate this with the discussion outlined in #243

**Comment by EthanMarx** [comment]
Thought about this a bit more and fleshed it out:

```python
@dataclass
class ImportanceSampler:
    """
    Class to calculate VT using importance sampling.

    Args:
        source: bilby PriorDict of the source distribution used to create injections
        recovered_parameters: array of recovered injections
        n_injections: number of total injections
        cosmology: astropy cosmology object used for volume calculation
    """
    source: "bilby.core.prior.PriorDict"
    recovered_parameters: np.ndarray
    n_injections: int
    cosmology: "Cosmology"
    
    def weights(self, target: "bilby.core.prior.PriorDict"):
        """
        Calculate the weights for the samples.
        """
        weights = target.prob(self.recovered_parameters) / self.source.prob(self.recovered_parameters)
        return weights

    def volume(self, cosmology: "Cosmology"):
        """
        Calculate the astrophysical volume over which injections have been made.
        """
        # module level function
        return calculate_astrophysical_volume(
            dl_min=self.source['luminosity_distance'].minimum,
            dl_max=self.source['luminosity_distance'].maximum,
            dec_min=self.source['dec'].minimum,
            dec_max=self.source['dec'].maximum,
            cosmology=cosmology
        )

    def calculate_vt(
        self, 
        target: "bilby.core.prior.PriorDict",
    ):
        
        """
        Calculates the VT using the target distribution for importance sampling
        """
        weights = self.weights(target)
        volume = self.volume(cosmology)
        weights *= volume
        return np.sum(weights) / self.n_injections
```

The idea here is that we instantiate the class with the source distribution, and then we can repeatedly call `calculate_vt`  for various `target` dists through

**Comment by alecgunny** [comment]
This is great. Couple qs:
1. Is `cosmology` meant to be a second argument to `calculate_vt`?
2. Should `target` be optional and default to `None`, in which case it will compute the `VT` under the source distribution?

**Comment by alecgunny** [comment]
(Or now that I'm re-reading it, should `self.volume` be an attribute that's computed and set during `__post_init__`, and then `self.volume` is used during `calculate_vt`?)

**Comment by EthanMarx** [comment]
Completely agree on all of these, had them implemented on my branch.

**Comment by EthanMarx** [comment]
closed with #252
